WalletWasabi/EventSourcing/InMemoryEventRepository.cs

```diff
@@ -0,0 +1,216 @@
+			}
+
+			// no action
+			Validated();
+
+			var newEvents = events.AddRange(wrappedEventsList);
+
+			// Atomically detect conflict and replace lastSequenceId and lock to ensure strong order in eventsBatches.
+			if (!aggregateEventsBatches.TryUpdate(
+				key: aggregateId,
+				newValue: (lastSequenceId, newEvents),
+				comparisonValue: (firstSequenceId - 1, events)))
+			{
+				Conflicted(); // no action
+				throw new OptimisticConcurrencyException($"Conflict while commiting events. Retry command. aggregate: '{aggregateType}' id: '{aggregateId}'");
+			}
+			Appended(); // no action
+
+			// If it is a first event for given aggregate.
+			if (tailSequenceId == 0)
+			{ // Add index of aggregate id into the dictionary.
+				IndexNewAggregateId(aggregateType, aggregateId);
+			}
+			return Task.CompletedTask;
+		}
+
+		public Task<IReadOnlyList<WrappedEvent>> ListEventsAsync(
+			string aggregateType,
+			string aggregateId,
+			long afterSequenceId = 0,
+			int? limit = null)
+		{
+			if (_aggregatesEventsBatches.TryGetValue(aggregateType, out var aggregateEventsBatches) &&
+				aggregateEventsBatches.TryGetValue(aggregateId, out var value))
+			{
+				var result = value.Events;
+				if (afterSequenceId > 0)
+				{
+					var foundIndex = result.BinarySearch(
+						new WrappedEvent(afterSequenceId),
+						Comparer<WrappedEvent>.Create((a, b) => a.SequenceId.CompareTo(b.SequenceId)));
+					if (foundIndex < 0)
+					{
+						foundIndex = ~foundIndex;
+					}
+					else
+					{
+						foundIndex++;
+					}
+					result = result.RemoveRange(0, foundIndex);
+				}
+				if (limit.HasValue && limit.Value < result.Count)
+				{
+					result = result.RemoveRange(limit.Value, result.Count - limit.Value);
+				}
+				return Task.FromResult((IReadOnlyList<WrappedEvent>)result);
+			}
+			return Task.FromResult(EmptyResult);
+		}
+
+		public Task<IReadOnlyList<string>> ListAggregateIdsAsync(
+			string aggregateType,
+			string? afterAggregateId = null,
+			int? limit = null)
+		{
+			if (_aggregatesIds.TryGetValue(aggregateType, out var tuple))
+			{
+				var ids = tuple.Ids;
+				var foundIndex = 0;
+				if (afterAggregateId != null)
+				{
+					foundIndex = ids.IndexOf(afterAggregateId);
+					if (foundIndex < 0)
+					{
+						foundIndex = ~foundIndex;
+					}
+					else
+					{
+						foundIndex++;
+					}
+				}
+				var result = new List<string>();
+				var afterLastIndex = limit.HasValue ?
+					Math.Min(foundIndex + limit.Value, ids.Count) :
+					ids.Count;
+				for (var i = foundIndex; i < afterLastIndex; i++)
+				{
+					result.Add(ids[i]);
+				}
+				return Task.FromResult((IReadOnlyList<string>)result.AsReadOnly());
+			}
+			return Task.FromResult(EmptyIds);
+		}
+
+		private void IndexNewAggregateId(string aggregateType, string id)
+		{
+			var tailIndex = 0L;
+			ImmutableSortedSet<string> aggregateIds;
+			ImmutableSortedSet<string> newAggregateIds;
+			var liveLockLimit = 10000;
+			do
+			{
+				if (liveLockLimit-- <= 0)
```

Comment:
I would put that `liveLockLimit--` on a separate line.

Reply:
done